### PR TITLE
Feature: Experiment Archive State

### PR DESCRIPTION
**Pull request by ppratikcr7**
This PR contains Archived status feature.

### CODE REVIEW

**Comment by danoswaltCL** [comment]
> This PR contains Archived status feature.

Can we get some more technical overview of the changes that were made to aid review and know how to test? If there is any finalized document that goes over the actual requirements that were decided on after all the years of discussion, can someone point to it?

**Comment by ppratikcr7** [comment]
> > This PR contains Archived status feature.
> 
> Can we get some more technical overview of the changes that were made to aid review and know how to test? If there is any finalized document that goes over the actual requirements that were decided on after all the years of discussion, can someone point to it?

Sure. So we do have a [document](https://docs.google.com/document/d/1X-DBcyz7nTwk_MLg8kYfWg1Uspb9bx0y9OEgi_T0tLI/edit#heading=h.1z8xhfd43vmo).

Requirements:

1. The archived experiments can't be unarchived or changed to a different status.
2. The experiment will be in read-only mode after being in archived state.
3. The logs wont get stored for archived experiments, even for the user earlier enrolled in it and had some logs.
4. The logs gets stored in a new Archived table as a backup.
5. While updating state to Archived state, we get a confirmation checkbox, until it's checked, the 'Save' button would be disable.
6. The archived experiments wont be shown on home dashboard experiment list view while the filter is 'All' or any other option except 'Status'. If we select 'Status' option, then we can filter all experiments including 'Archived' using a mat-auto complete search input box.

Just wanted to confirm one thing, we don't want to have archived state when the experiment is in enrolling state as per the diagram? And wait for experiment to be in enrolment Complete state.

213239091-10d9c46f-04d5-4d27-b08e-9f77299a42f9

**Review by bcb37 — COMMENTED** [review]
I've gone over this and it all looks good to me.

**Comment by ppratikcr7** [comment]
> I've gone over this and it all looks good to me.

Okay, you can also approve this and it would't auto merge as I have disabled the option as Dan had also pointed out in slack. We can wait for multiple approvals before merging if needed.

File: backend/packages/Upgrade/src/api/models/ArchivedStats.ts
Comment (danoswaltCL) [inline backend/packages/Upgrade/src/api/models/ArchivedStats.ts:13]: sry gotta be annoying about `any`, is there any type that can be defined for this? What does "result" mean?

File: frontend/projects/upgrade/src/app/features/dashboard/home/components/experiment-list/experiment-list.component.ts
Comment (danoswaltCL) [inline frontend/projects/upgrade/src/app/features/dashboard/home/components/experiment-list/experiment-list.component.ts:59]: we really only want searchValue to specifically be a string, so I think we'd be better to be explicit `if (typeof this.searchValue === 'string')`.

File: frontend/projects/upgrade/src/app/features/dashboard/home/components/experiment-list/experiment-list.component.ts
Comment (danoswaltCL) [inline frontend/projects/upgrade/src/app/features/dashboard/home/components/experiment-list/experiment-list.component.ts:145]: same as other comment, filterValue needs to be a string, not just 'not undefined', a null value can slip in here, for instance.

File: frontend/projects/upgrade/src/assets/i18n/en.json
Comment (danoswaltCL) [inline frontend/projects/upgrade/src/assets/i18n/en.json:271]: probably should have period at the end.

**Comment by danoswaltCL** [comment]
Just a few small comments from me, there's just the one spot in `experiment-list` with the filters that feels like it could be refactored to be simpler, but as I looked at it I didn't have any good quick suggestions. Chatgpt also didn't give me any good suggestions.

File: backend/packages/Upgrade/src/api/controllers/QueryController.ts
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/controllers/QueryController.ts:53]: Why do we need a separate http endpoint for archive? I thing We should use the /analyse endpoint and server should get the data from the archive table in case the experiment is in archive state. Given that we have the same response type we can use the same endpoint. This will also keep the frontend dumb.

File: backend/packages/Upgrade/src/api/repositories/MetricRepository.ts
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/repositories/MetricRepository.ts:33]: Changes over here are unnecessary. Because the query does more than what the function name suggests.

File: frontend/projects/upgrade/src/app/core/analysis/store/analysis.actions.ts
Comment (VivekFitkariwala) [inline frontend/projects/upgrade/src/app/core/analysis/store/analysis.actions.ts:37]: You don't need to send the state information from the frontend. You can get that from the queryId itself.

**Review by VivekFitkariwala — CHANGES_REQUESTED** [review]
The frontend doesn't need to know that we are storing the query result of the archive state in a different way. We can use the same endpoint to serve the result.

**Comment by RidhamShah** [comment]
@VivekFitkariwala  Yes, we can merge analysis and achieve branch by giving one more variable to identify which call is sent.

File: backend/packages/Upgrade/src/api/models/ArchivedStats.ts
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/models/ArchivedStats.ts:13]: the result stores the query result in a JSON object. Will update this to appropriate type

File: frontend/projects/upgrade/src/app/core/analysis/store/analysis.actions.ts
Comment (RidhamShah) [reply frontend/projects/upgrade/src/app/core/analysis/store/analysis.actions.ts]: code updated

File: backend/packages/Upgrade/src/api/controllers/QueryController.ts
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/controllers/QueryController.ts]: code updated

File: backend/packages/Upgrade/src/api/models/ArchivedStats.ts
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/models/ArchivedStats.ts]: code updated

File: frontend/projects/upgrade/src/app/features/dashboard/home/components/experiment-list/experiment-list.component.ts
Comment (ppratikcr7) [reply frontend/projects/upgrade/src/app/features/dashboard/home/components/experiment-list/experiment-list.component.ts:147]: yeah this was a very old code which used undefined. Nice catch, will update. :)

File: backend/packages/Upgrade/src/api/services/QueryService.ts
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/services/QueryService.ts:72]: You don't need to `await` here. Just return the promise

**Comment by VivekFitkariwala** [comment]
The changes look good.

File: backend/packages/Upgrade/src/api/services/QueryService.ts
Comment (ppratikcr7) [reply backend/packages/Upgrade/src/api/services/QueryService.ts:72]: @RidhamShah Can you please resolve this.

**Review by VivekFitkariwala — APPROVED** [review]

**Pull request merged**
